### PR TITLE
[WIP] Adds IPC as a species

### PR DESCRIPTION
**Pull request by SgtHunk**
## About The Pull Request

This adds the Integrated Positronic Chassis species to the game, ported from various PRs (Mainly Skyrat). It adds support  for limb overlays (which beefmen now use to avoid editing that bodypart TG file a lot).
TV-head robots powered by Positronic brains. Pretty fun. They're completely machines, with robotic bodyparts and stuff. 

image

## Why It's Good For The Game

More species (whilst editing as little TG files as possible). A lot of people enjoyed the thought of a robotic species, and besides, I'm sure a LOT of people want IPC.

## Changelog
:cl:
add: Added IPC as a species
/:cl:

### CODE REVIEW

**Comment by remove32** [comment]
idk bout this

fulp was never intended as a server for the kind of RP that this inspires
IPCs kinda run counter to the direction we (swain and I) discussed about where we want to take the server
keeping silicons restrained in some way is pretty necessary to lore and balance

plus they just look like robotic furries and i do not vibe with that.

discussion welcome

**Comment by JohnFulpWillard** [comment]
IPCs arent meant to replace cyborgs, just as another playable race like Ethereal/Plasmaman. All the costumes (except the one at the very top right) all seem perfectly fine, and I'm pretty sure are from Beestation, a server similar to ours.

**Comment by remove32** [comment]
tell you what. get them working, prove they work to me, and i will talk to swain and if swain is ok with it i'll run a poll

**Comment by SgtHunk** [comment]
This PR is done. Tested it, IPCs are fully functional. Open to feedback.
Only need to change a bit on the species itself (removing commented code, and unnecessary modifiers, also making them weaker to damage)
I'll try to find another sprite for the Pharmaceutics IPC chassis (The top right pink/white model).

**Comment by EIB0N** [comment]
Im down to have IPC's in as long as I get to see the windows xp version among them

**Comment by John-Dang** [comment]
I've been thinking that a synthetic/android species would be great to add, and this is just perfect. Some of the sprites maybe not but overall very good. I hope all their internal organs are also synthetic ones, while still needing to recharge like ethereals do?

**Comment by BootlegBow** [comment]
How will they work with quirks? A lot of quirks won't really be able to work with IPCs because of a lot of reasons like not being able to digest chems

**Comment by SgtHunk** [comment]
> How will they work with quirks? A lot of quirks won't really be able to work with IPCs because of a lot of reasons like not being able to digest chems

That is, in fact, an issue.
The original PR added flags for reagents whether they were synthetic/organic. Alongside synthetic only chemicals. I can probably figure out something.

File: code/modules/surgery/bodyparts/_bodyparts.dm [hunk 1]
```diff
@@ -829,25 +835,15 @@
 
 	if(is_organic_limb())
 		if(should_draw_greyscale)
-//			limb.icon = 'icons/mob/human_parts_greyscale.dmi'
-			switch(species_id) // [FULP EDIT STARTS]
-				if("beefman") //If we ever add more unique races - this should be a list of them.
-					limb.icon = 'fulp_modules/beefman_port/icons/mob/beefman_bodyparts.dmi'
-				else
-					limb.icon = 'icons/mob/human_parts_greyscale.dmi'// [FULP EDIT END]
+			limb.icon = 'icons/mob/human_parts_greyscale.dmi'
```
Comment (GoldenAlpharex) [inline code/modules/surgery/bodyparts/_bodyparts.dm:838]: Why did you remove that? Did you implement it differently?

Comment (SgtHunk) [reply code/modules/surgery/bodyparts/_bodyparts.dm:838]: Overwritten in ZFulpcode, _bodyparts.dm

File: fulp_modules/ZFulpcode/modules/ipc/objective.dm [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+				possible_targets -= PT
+		if(!possible_targets.len)
+			possible_targets = all_possible_targets
+	if(possible_targets.len > 0)
+		target = pick(possible_targets)
+	update_explanation_text()
+	return target
+
+/datum/objective/proc/is_valid_target(possible_target)
+	return TRUE
```
Comment (ATH1909) [inline fulp_modules/ZFulpcode/modules/ipc/objective.dm:28]: why is this in this folder/pr?

File: fulp_modules/ipc/code/human_defense.dm [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	var/affects_leg = FALSE
+	var/stun_time = 0
+	for(var/obj/item/bodypart/L in src.bodyparts)
+		if(L.status == BODYPART_ROBOTIC)
+			if(!informed)
+				to_chat(src, "<span class='userdanger'>You feel a sharp pain as your robotic limbs overload.</span>")
+				informed = TRUE
+			switch(severity)
+				if(1)
+					L.receive_damage(0,6)
```
Comment (ATH1909) [inline fulp_modules/ipc/code/human_defense.dm:15]: why is the damage for this so low?

File: fulp_modules/ipc/code/human_defense.dm [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+					stun_time += 40
+				if(2)
+					L.receive_damage(0,3)
```
Comment (ATH1909) [inline fulp_modules/ipc/code/human_defense.dm:18]: does this damage pierce the damage reduction of augged limbs, or does it just get no-selled by them?

File: fulp_modules/chemistry/other_reagents.dm [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+/datum/reagent/fuel/on_mob_life(mob/living/carbon/C)
+	if(C.mob_biotypes & MOB_ROBOTIC)
+		C.nutrition = min(C.nutrition + 5, NUTRITION_LEVEL_FULL-1)
+	..()
+
+/datum/reagent/oil/on_mob_life(mob/living/carbon/C)
+	if(C.mob_biotypes & MOB_ROBOTIC && C.blood_volume < BLOOD_VOLUME_NORMAL)
+		C.blood_volume += 0.5
+	..()
```
Comment (ATH1909) [inline fulp_modules/chemistry/other_reagents.dm:35]: note that these chems, along with plasma, will still try to deal tox damage to robotic mobs they're in. ipcs are tox immune, but if you add another robotic race in the future that isn't, this could lead to some issues.

File: fulp_modules/ipc/code/ipc.dm [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	H.update_body()
+	H.say("Reactivating [pick("core systems", "central subroutines", "key functions")]...")
+	sleep(3 SECONDS)
+	H.say("Reinitializing [pick("personality matrix", "behavior logic", "morality subsystems")]...")
+	sleep(3 SECONDS)
+	H.say("Finalizing setup...")
+	sleep(3 SECONDS)
+	H.say("Unit [H.real_name] is fully functional. Have a nice day.")
+	H.dna.features["ipc_screen"] = saved_screen
+	H.update_body()
```
Comment (ATH1909) [inline fulp_modules/ipc/code/ipc.dm:52]: will this be spammed if you repeatedly meet and unmeet the conditions needed to revive?

I don't like the sleeps here...

File: fulp_modules/ipc/code/ipc.dm [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	default_color = "00FF00"
+	inherent_biotypes = MOB_ROBOTIC|MOB_HUMANOID
+	inherent_traits = list(TRAIT_ADVANCEDTOOLUSER,TRAIT_RADIMMUNE,TRAIT_VIRUSIMMUNE,TRAIT_NOBREATH,TRAIT_TOXIMMUNE,TRAIT_NOCLONELOSS,TRAIT_GENELESS,TRAIT_STABLEHEART,TRAIT_NO_HUSK)
+	species_traits = list(MUTCOLORS_PARTSONLY,LIPS,NOEYESPRITES,NOTRANSSTING,REVIVES_BY_HEALING,ROBOTIC_LIMBS,TRAIT_RESISTCOLD,NOZOMBIE,TRAIT_PIERCEIMMUNE, TRAIT_NOFLASH)
+	mutant_bodyparts = list("ipc_screen" = "BSOD", "ipc_antenna" = "None", "ipc_chassis" = "Morpheus Cyberkinetics(Greyscale)")
+	changesource_flags = MIRROR_BADMIN | WABBAJACK | MIRROR_MAGIC | MIRROR_PRIDE | ERT_SPAWN | RACE_SWAP | SLIME_EXTRACT
+	burnmod = 1.5
+	heatmod = 1.2
+	brutemod = 1.5
+	siemens_coeff = 1.2 //Not more because some shocks will outright crit you, which is very unfun
```
Comment (ATH1909) [inline fulp_modules/ipc/code/ipc.dm:14]: you accounted for this stacking with burnmod, right?

File: fulp_modules/chemistry/other_reagents.dm [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+/datum/reagent/fuel/on_mob_life(mob/living/carbon/C)
+	if(C.mob_biotypes & MOB_ROBOTIC)
+		C.nutrition = min(C.nutrition + 5, NUTRITION_LEVEL_FULL-1)
+	..()
+
+/datum/reagent/oil/on_mob_life(mob/living/carbon/C)
+	if(C.mob_biotypes & MOB_ROBOTIC && C.blood_volume < BLOOD_VOLUME_NORMAL)
+		C.blood_volume += 0.5
```
Comment (ATH1909) [inline fulp_modules/chemistry/other_reagents.dm:34]: isn't oil already the exotic_blood of ipcs?

File: fulp_modules/ipc/code/organs/ears.dm [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	. = ..()
+	if(!owner || . & EMP_PROTECT_SELF)
+		return
+	switch(severity)
+		if(1)
+			owner.Jitter(30)
+			owner.Dizzy(30)
+			owner.Knockdown(80)
+			deaf = 30
+			to_chat(owner, "<span class='warning'>Your robotic ears are ringing, uselessly.</span>")
```
Comment (ATH1909) [inline fulp_modules/ipc/code/organs/ears.dm:22]:
```suggestion
			to_chat(owner, "<span class='warning'>Your robotic ears are ringing uselessly.</span>")
```

File: fulp_modules/ipc/code/organs/liver.dm [hunk 10]
```diff
@@ -0,0 +1,8 @@
+/obj/item/organ/liver/robot_ipc
+	name = "reagent processing unit"
+	desc = "An electronic device that processes the beneficial chemicals for the synthetic user."
+	organ_flags = ORGAN_SYNTHETIC
+	status = ORGAN_ROBOTIC
+	icon = 'fulp_modules/ipc/icons/ipc_organs.dmi'
+	icon_state = "liver-c"
+	filterToxins = FALSE //We dont filter them, we're immune ot them
```
Comment (ATH1909) [inline fulp_modules/ipc/code/organs/liver.dm:8]:
```suggestion
	filterToxins = FALSE //We dont filter them, we're immune to them
```

File: fulp_modules/ipc/code/organs/lungs.dm [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	safe_oxygen_min = 0	//What are you doing man, dont breathe with those!
+	safe_oxygen_max = 0
+	cold_level_1_damage = 0
+	cold_level_2_damage = 0
+	cold_level_3_damage = 0
+	status = ORGAN_ROBOTIC
+	organ_flags = ORGAN_SYNTHETIC
+	slot = ORGAN_SLOT_LUNGS
+
+/obj/item/organ/lungs/robot_ipc/emp_act(severity)
```
Comment (ATH1909) [inline fulp_modules/ipc/code/organs/lungs.dm:21]: do TRAIT_NOBREATH species even need lungs? if they don't, then wouldn't it be beneficial for an IPC to get their lungs removed at roundstart so that they don't overheat when EMP'd?

File: fulp_modules/ipc/code/organs/stomach.dm [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+/obj/item/organ/stomach/robot_ipc/emp_act(severity)
+	. = ..()
+	if(!owner || . & EMP_PROTECT_SELF)
+		return
+	switch(severity)
+		if(1)
+			owner.nutrition = 50
+			to_chat(owner, "<span class='warning'>Alert: Detected severe battery discharge!</span>")
+		if(2)
+			owner.nutrition = 250
```
Comment (ATH1909) [inline fulp_modules/ipc/code/organs/stomach.dm:21]: does this mean that you can EMP yourself to gain nutrition if you have less than 250 nutrition left?

File: fulp_modules/ipc/code/organs/stomach.dm [hunk 13]
```diff
@@ -0,0 +1,22 @@
+/obj/item/organ/stomach/robot_ipc
+	name = "IPC micro cell"
+	icon = 'fulp_modules/ipc/icons/ipc_organs.dmi'
+	icon_state = "stomach-ipc"
+	w_class = WEIGHT_CLASS_NORMAL
```
Comment (ATH1909) [inline fulp_modules/ipc/code/organs/stomach.dm:5]: >micro cell
>WEIGHT_CLASS_NORMAL

bruh moment

Comment (ATH1909) [reply fulp_modules/ipc/code/organs/stomach.dm:5]: why is a human-sized creature running on a MICRO cell, anyway? shouldn't it be a MACRO cell or something?

File: fulp_modules/ipc/code/surgery/body_repair.dm [hunk 14]
```diff
@@ -0,0 +1,128 @@
+//Almost copypaste of tend wounds, with some changes
+/datum/surgery/robot_healing
+	steps = list(/datum/surgery_step/mechanic_open,
+				/datum/surgery_step/pry_off_plating,
+				/datum/surgery_step/cut_wires,
+				/datum/surgery_step/robot_heal,
+				/datum/surgery_step/mechanic_close)
+
+	requires_bodypart_type = BODYPART_ROBOTIC
```
Comment (ATH1909) [inline fulp_modules/ipc/code/surgery/body_repair.dm:9]: is it intended that, unlike robotic!organ manipulation, you can't do this surgery on yourself?

File: fulp_modules/ipc/code/surgery/body_repair.dm [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+	desc = "A surgical procedure that provides repairs and maintenance to robotic limbs. Is slightly more efficient when the patient is severely damaged."
+	replaced_by = null
+
+/***************************STEPS***************************/
+
+/datum/surgery_step/robot_heal/basic
+	name = "repair damage"
+	brutehealing = 10
+	burnhealing = 10
+	missinghpbonus = 15
```
Comment (ATH1909) [inline fulp_modules/ipc/code/surgery/body_repair.dm:128]: will this show up alongside the abstract version of this surgery?

also, why would I ever do this surgery when just directly applying a welder or cable coils would heal an IPC faster and more efficiently (IIRC) and there are no ghetto alternatives to using a welder or cable coils?

File: fulp_modules/ipc/code/surgery/chassis_restoration.dm [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+	/datum/surgery_step/weld_plating/fullbody,
+	/datum/surgery_step/finalize_chassis_restoration)
+
+	requires_bodypart_type = BODYPART_ROBOTIC
+	target_mobtypes = list(/mob/living/carbon/human)
+	possible_locs = list(BODY_ZONE_CHEST)
+	desc = "A surgical procedure that rebuilds a synthetic unit from their skeleton to full integrity, recommended if they are damaged far beyond repair."
+
+/datum/surgery_step/pry_off_plating/fullbody
+	time = 120
```
Comment (ATH1909) [inline fulp_modules/ipc/code/surgery/chassis_restoration.dm:19]: holy shit this takes 12 seconds for each step (except a few of them)???

File: fulp_modules/ipc/code/surgery/restore_chassis_looks.dm [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+	requires_bodypart_type = BODYPART_ROBOTIC
+	possible_locs = list(BODY_ZONE_CHEST)
+	desc = "A procedure that welds the robotic limbs back into the patient's preferred state aswell as re-applying their paintjob."
+
+/datum/surgery_step/restore_paintjob
+	name = "spray paint"
+	implements = list(
+		/obj/item/toy/crayon/spraycan = 100)
+	time = 58
```
Comment (ATH1909) [inline fulp_modules/ipc/code/surgery/restore_chassis_looks.dm:15]: let apcs do these surgeries on themselves

Comment (ATH1909) [reply fulp_modules/ipc/code/surgery/restore_chassis_looks.dm:15]: or at least let the spray paint one be used on yourself; it really shouldn't even be a surgery, tbh

File: fulp_modules/ipc/code/surgery/mechanic_steps.dm [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+	if(plat.get_amount() < metalamount)
+		to_chat(user, "<span class='warning'>Not enough metal!</span>")
+		return FALSE
+	return TRUE
+
+/datum/surgery_step/add_plating/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	var/obj/item/stack/sheet/metal/plat = tool
+	if(plat && !(plat.get_amount()<metalamount)) //failproof
+		plat.use(metalamount)
+	return TRUE
```
Comment (ATH1909) [inline fulp_modules/ipc/code/surgery/mechanic_steps.dm:98]: I think a thing you can do here is start the plating adding process, then use alt+click (perhaps even an alt+click menu that you brought up and typed in before even beginning the step) to remove enough sheets from the stack in your active hand to let you avoid needing to spend any metal upon completing the step

File: code/modules/surgery/bodyparts/_bodyparts.dm [hunk 19]
```diff
@@ -509,6 +509,11 @@
 			update_disabled()
 		if(updating_health)
 			owner.updatehealth()
+		// FULP EDIT BEGIN
+		if(owner.stat == DEAD && owner?.dna?.species && (REVIVES_BY_HEALING in owner.dna.species.species_traits))
+			if(owner.health > 50)
```
Comment (Fikou) [inline code/modules/surgery/bodyparts/_bodyparts.dm:514]: this should be a define

File: code/__DEFINES/_Fulpdefines/ipc.dm [hunk 20]
```diff
@@ -0,0 +1,4 @@
+#define TRAIT_NO_HUSK				"no_husk" // If we have this trait, we can't get husked.
+#define REVIVES_BY_HEALING				33 // Used to determine if we revive after our health reaches a certain threshold.
```
Comment (Fikou) [inline code/__DEFINES/_Fulpdefines/ipc.dm:2]: maybe mention what the threshold is?

File: fulp_modules/ipc/code/status_procs.dm [hunk 21]
```diff
@@ -0,0 +1,11 @@
+/mob/living/carbon/human/become_husk(source)
+	if(!HAS_TRAIT(src, TRAIT_NO_HUSK))
```
Comment (Fikou) [inline fulp_modules/ipc/code/status_procs.dm:2]: add this trait to skeletons and remove that hard typecheck

File: fulp_modules/ipc/code/surgery/body_repair.dm (hunk 14 above)
Comment (Fikou) [reply fulp_modules/ipc/code/surgery/body_repair.dm:9]: lol

File: fulp_modules/ipc/code/surgery/chassis_restoration.dm [hunk 22]
```diff
@@ -0,0 +1,77 @@
+/datum/surgery/robot_chassis_restoration
+	name = "Full chassis restoration (Full heal + revival)"
```
Comment (Fikou) [inline fulp_modules/ipc/code/surgery/chassis_restoration.dm:2]: restoration should be enough for people to realize it revives, no need to make it weirdly ocky

File: fulp_modules/ipc/code/surgery/chassis_restoration.dm [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+/datum/surgery_step/finalize_chassis_restoration/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	display_results(user, target, "<span class='notice'>You begin to finalize [target]'s chassis...</span>",
+			"[user] begins to finalize [target]'s chassis.",
+			"[user] begins to finalize [target]'s chassis.")
+
+/datum/surgery_step/finalize_chassis_restoration/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	target.cure_husk()
+	target.heal_overall_damage(1000,1000,0, BODYPART_ROBOTIC)
+	//I call this twice because for some reason health may not update and we may not get a revival, and I cant find a good clean solution to this
```
Comment (Fikou) [inline fulp_modules/ipc/code/surgery/chassis_restoration.dm:74]: then call updatehealth() or something not this horrific thing

File: fulp_modules/ipc/code/surgery/chassis_restoration.dm [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+	time = 120
+
+/datum/surgery_step/finalize_chassis_restoration/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	display_results(user, target, "<span class='notice'>You begin to finalize [target]'s chassis...</span>",
+			"[user] begins to finalize [target]'s chassis.",
+			"[user] begins to finalize [target]'s chassis.")
+
+/datum/surgery_step/finalize_chassis_restoration/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	target.cure_husk()
+	target.heal_overall_damage(1000,1000,0, BODYPART_ROBOTIC)
```
Comment (Fikou) [inline fulp_modules/ipc/code/surgery/chassis_restoration.dm:73]: use INFINITY if you want ultra epic heal, this just looks like ugly magic number

File: fulp_modules/ipc/code/surgery/chassis_restoration.dm [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+/datum/surgery_step/pry_off_plating/fullbody
+	time = 120
+
+/datum/surgery_step/pry_off_plating/fullbody/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	display_results(user, target, "<span class='notice'>You begin to pry off [target]'s plating...</span>",
+			"[user] begins to pry off [target]'s plating.",
+			"[user] begins to pry off [target]'s plating.")
+
+/datum/surgery_step/cut_wires/fullbody
+	time = 120
```
Comment (Fikou) [inline fulp_modules/ipc/code/surgery/chassis_restoration.dm:27]: use second defines for all of these
```suggestion
	time = 12 SECONDS
```

File: fulp_modules/ipc/code/ipc.dm (hunk 6 above)
Comment (remove32) [reply fulp_modules/ipc/code/ipc.dm:52]: sleeps in general are bad practice

**Comment by SgtHunk** [comment]
haha fuck this new branch it is
